airbyte-integrations/connectors/source-blob/source_blob/fileclient.py

```diff
@@ -0,0 +1,263 @@
+        #         uri = f"{storage}{user}@{host}:{port}/{url}"
+        #     return smart_open.open(uri, transport_params=transport_params, mode=mode)
+        # return smart_open.open(self.fullurl, mode=mode)
+
+
+class FileClientS3(FileClient):
+    """TODO docstring"""
+
+    class _Decorators():
+        """ TODO docstring """
+        @classmethod
+        def init_boto_session(cls, func):
+            """TODO Docstring """
+            def inner(self, *args, **kwargs):
+                # why we're making a new Session at file level rather than stream level
+                # https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html#multithreading-and-multiprocessing
+                if not hasattr(self, '_boto_session'):
+                    if self.use_aws_account:
+                        self._boto_session = boto3session.Session(
+                            aws_access_key_id=self._provider.get("aws_access_key_id"),
+                            aws_secret_access_key=self._provider.get("aws_secret_access_key"))
+                    else:
+                        self._boto_session = boto3session.Session()
+                if not hasattr(self, '_boto_s3_resource'):
+                    if self.use_aws_account:
+                        self._boto_s3_resource = self._boto_session.resource('s3')
+                    else:
+                        self._boto_s3_resource = self._boto_session.resource('s3', config=Config(signature_version=UNSIGNED))
+                return func(self, *args, **kwargs)
+            return inner
+
+    @property
+    def storage_scheme(self) -> str:
+        return "s3://"
+
+    @property
+    def url(self) -> str:
+        bucket = self._provider.get("bucket")
+        return f"{bucket}/{self._url}"
+
+    @property
+    @_Decorators.init_boto_session
+    def last_modified(self) -> datetime:
+        """ TODO docstring """
+        obj = self._boto_s3_resource.Object(self._provider.get('bucket'), self._url)
+        return obj.last_modified
+
+    @staticmethod
+    def use_aws_account(provider: dict) -> bool:
+        aws_access_key_id = provider.get("aws_access_key_id")
+        aws_secret_access_key = provider.get("aws_secret_access_key")
+        return aws_access_key_id and aws_secret_access_key
+
+    def _open(self, binary):
+        """ TODO docstring """
+        mode = "rb" if binary else "r"
+        if self.use_aws_account(self._provider):
+            aws_access_key_id = self._provider.get("aws_access_key_id", "")
+            aws_secret_access_key = self._provider.get("aws_secret_access_key", "")
+            result = smart_open.open(f"{self.storage_scheme}{aws_access_key_id}:{aws_secret_access_key}@{self.url}", mode=mode)
+        else:
+            config = Config(signature_version=UNSIGNED)
+            params = {
+                "resource_kwargs": {"config": config},
+            }
+            result = smart_open.open(self.full_url, transport_params=params, mode=mode)
```

Comment:
just confirming, this reads the file in a streaming fashion right? i.e doesn't load the whole thing into memory?

Reply:
Yeah it returns a file-like object (I/O stream)